### ISSUE TITLE
HOMEDRIVE is D: but $HOME and ~ still use C:

### ISSUE THREAD

**Issue by TonyStachnicki** [PowerShell/PowerShell#17685]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://docs.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

Starting with a user profile / home directory called C:\Users\tonys
Create a separate home directory called D:\Users\tonys
Move unhidden folders into the new folder, using Microsoft-approved methods.
Run command: net user tonys /homedir:D:\Users\tonys
Sign-off/sign-on.
The values of %HOMEDRIVE% and $ENV:HOMEDRIVE are now "D: ".
But in PowerShell the values of $HOME and the "~" character are still "C:\Users\tonys".

I have reproduced this using .zip versions 6.2.4, 7.0.0, 7.2.5, and 7.3.0 preview 5.

The documentation for the Get-PSProvider cmdlet defines the Home property of the FileSystem provider as  $env:homedrive\$env:homepath or $home.
Link: https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.management/get-psprovider?view=powershell-7.2

### Expected behavior

```console
PS C:\Users\tonys> $ENV:HOMEDRIVE
D:
PS C:\Users\tonys> $HOME
D:\Users\tonys
PS C:\Users\tonys> (Resolve-Path ~).Path
D:\Users\tonys
```

### Actual behavior

```console
PS C:\Users\tonys> $ENV:HOMEDRIVE
D:
PS C:\Users\tonys> $HOME
C:\Users\tonys
PS C:\Users\tonys> (Resolve-Path ~).Path
C:\Users\tonys
```

### Environment data

```powershell
Name                           Value
----                           -----
PSVersion                      7.2.5
PSEdition                      Core
GitCommitId                    7.2.5
OS                             Microsoft Windows 10.0.22000
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

**Comment by jhoneill** [comment]
What do  `[System.Environment]::GetFolderPath([System.Environment+SpecialFolder]::UserProfile)` and `$env:UserProfile` return

**Comment by mklement0** [comment]
Unfortunately, PowerShell  on Windows has always used the value of  `$env:USERPROFILE` as `$HOME` and `~` (in the `FileSystem` provider), which is only _by default_ the same as  `${env:HOMEDRIVE}${env:HOMEPATH}`

If the home directory is configured to differ from the profile location, you'll have to use the latter, at least currently.

**Comment by TonyStachnicki** [comment]
<kbd>[System.Environment]::GetFolderPath([System.Environment+SpecialFolder]::UserProfile)</kbd> and <kbd>$env:UserProfile</kbd> both return "C:\Users\tonys".

**Comment by TonyStachnicki** [comment]
It seems to me that the user profile and the home directory are two logically separate folders that share the same physical folder in Windows. Is it at all likely that <kbd>$HOME</kbd> will be changed to <kbd>$env:HOMEDRIVE$env:HOMEPATH</kbd> in the near future, or ever?

**Comment by TonyStachnicki** [comment]
[System.Environment]::GetFolderPath([System.Environment+SpecialFolder]::UserProfile)
 and $env:UserProfile both return "C:\Users\tonys".

**Comment by jhoneill** [comment]
> It seems to me that the user profile and the home directory are two logically separate folders that share the same physical folder in Windows. Is it at all likely that $HOME will be changed to $env:HOMEDRIVE$env:HOMEPATH in the near future, or ever?

There **are** two functions - IIRC profile was originally the home of NTUser.Dat - and is set for each user in the registry,    
and the home directory was where users put personal files. The two are moved independently if  `net user tonys /profilepath:newDir` is combined with the `/homedir`  version everything is in one places otherwise you get the functions delivered from different directories.  

When the two are different, whether ~ should be home or profile is arguable both ways but changing it will inevitably break things.   
I did a scan for things that use $home and found that most of the use is combined with \Documents (and in a few cases with \Pictures) when the script should use `[System.Environment]::GetFolderPath()` these scripts will fail when `Documents` etc are moved under `onedrive` for example (I have one machine set that way). 
 Having $home as  `%HomeDrive%%HomeDir%`  instead of %userProfile% would mean these scripts have a chance of working when home <> profile.  
 So I can see that $home **MIGHT** be changed.  How likely or how soon ? That depends on persuading people that what it fixes is worth fixing and doing so has little chance of side effects. That probably needs a bigger scan of how it is used.

**Comment by TonyStachnicki** [comment]
I have run my PCs with the "Known" and OneDrive folders in folder
D:\Users\tonys for many years.
I use a PowerShell profile script to set up each session with what I
consider to be the correct $HOME variable and have been happy with the
results.
I find this folder arrangement is very good for backup and recovery.
I can restore Windows partitions without upsetting my data partition and
vice-versa.

A few weeks ago I discovered that I could change the HOMEDRIVE using the
"net user" command.
This works for the Command Prompt, the working directory is now
"D:\Users\tonys".
And for expanded string properties in the registry, %HOMEDRIVE% is replaced
by "D:".
I was disappointed to find that although PowerShell has $Env:HOMEDRIVE set
to "D:", the $HOME variable has not changed.

However I can understand that the benefits of changing $HOME to use the
HOMEDRIVE are small and many other users might experience problems.

**Comment by jhoneill** [comment]
Is there a reason why you have left the profile on C: and not used the second `net user` command to move it to D: as well ? 

It's a minority of people who split the drive into multiple partitions, and of those only the ones who put home on one and keep profile on the other are going to hit this issue.  For everyone else it doesn't matter if $home is set homedrive\homepath, or userprofile. It's been the latter (wrongly) for ever.  For the few affected something will be fixed if it is changed. But $home isn't used by many things, and where it is used it's not reliable. The advice really should be not to rely on $home but where things do, life can be made better for a few people by changing. So can you provide a list of things that aren't working because $home is wrong ? Because if it's just an annoyance that an unused variable has the wrong value in some "edge cases" it's hard to see it getting attention.

**Comment by TonyStachnicki** [comment]
I like the fact that all the software is on the "C" volume and it's easy to
remove changes by restoring this volume, without affecting my personal
files on the "D" volume.
The AppData folder in "C:\Users\tonys" contains application packages.
The registry hive files in "C:\Users\tonys" contain software configuration
settings.
So these programs and settings stay in sync with the operating system and
the other application software.

When I reinstall Windows, I choose to completely replace the old system, so
everything on the "C" volume is reset to a clean state.
I have a PowerShell script that uses the "KnownFolder" programming
interface functions to change the folder paths of the relocated folders to
the appropriate "D:\Users\tonys" folders.
I have never had any problems with this approach.

For years I have used a PowerShell profile script to set $HOME and the
initial working directory to my preferred location, so I don't have to hard
code values in the rest of my code.
I will just carry on using this script.

**Comment by mklement0** [comment]
As for the importance of `$HOME` and its if-the-current-location-is-a-file-system-location equivalent, `~`:

It's an invaluable, cross-platform way to refer to the current user's home directory, and, at least from what I can tell, frequently used - but it has inherent challenges:

Given the variable's name:

* From a Windows-only perspective (which applies to  _Windows PowerShell_), it should always have been the equivalent of `%HOMEDRIVE%HOMEPATH%` (expressed in `cmd.exe` terms), not the value of `%USERPROFILE%`, which only _by default_ is the same as the former.
  * The distinction between the two location is where _data_ (home) is stored  vs. where _configuration_ (profile) is stored.
  * As an aside: Arguably, _Windows itself_ should provide a _single_ environment variable that reflects the value of `%HOMEDRIVE%HOMEPATH%`

* On Unix-like platforms, there's no such distinction, and `$HOME` reliably refers to the user's home directory.

Given the _cross-platform_ nature of PowerShell Core, there is an inherent tension:

* The _current_ definition of `$HOME` reliably refers to where _some_ applications - notably those with a Unix heritage or those designed with multi-platform support - store their _configuration_ information (e.g, `~/.dotnet`)

* On Unix-like platforms, it also _reliably_ refers to where a user's data (documents) are stored, but on Windows it does so only _situationally_, depending on the host machine's configuration.

In short: To resolve this tension and to account for other platform differences , true _platform-agnostic abstractions_ are needed - see #6966
